File: CulinarySite/ServiceLayer/TagService.cs
using Database;
using Repositories;
using System.Collections.Generic;
using ServiceLayer.ViewModels.Tag;

namespace ServiceLayer
{
    public class TagService : ITagService
    {
        private readonly IReadOnlyGenericRepository<Tag> tagReadOnlyRepository;
        private readonly IWriteGenericRepository<Tag> tagWriteRepository;
        public TagService(
            IReadOnlyGenericRepository<Tag> tagReadOnlyRepository,
            IWriteGenericRepository<Tag> tagWriteRepository)
        {
            this.tagReadOnlyRepository = tagReadOnlyRepository;
            this.tagWriteRepository = tagWriteRepository;
        }

        public void CreateTag(CreateTagModel createTagModel)
        {
            var tag = new Tag
            {
                Text = createTagModel.Text
            };
            this.tagWriteRepository.Create(tag);
            this.tagWriteRepository.Save();
        }

        public void UpdateTag(UpdateTagModel updateTagModel)
        {
            var tag = new Tag
            {
                Id = updateTagModel.TagId,
                Text = updateTagModel.Text
            };
            this.tagWriteRepository.Update(tag);
            this.tagWriteRepository.Save();
        }

        public void DeleteTag(int id)
        {
            this.tagWriteRepository.Delete(id);
            this.tagWriteRepository.Save();
        }

        public IEnumerable<TagListModel> GetTagList(bool withRelated)
        {
            IEnumerable<Tag> tags;
            List<TagListModel> tagListModels = new List<TagListModel>();
            if (withRelated)
            {
                tags = this.tagReadOnlyRepository.GetItemListWithInclude(
                                x => x.Recipes,
                                x => x.Episodes);
                foreach (var tag in tags)
                {
                    tagListModels.Add(new TagListModel
                    {
                        TagId = tag.Id,
                        Text = tag.Text
                    });
                }
                return tagListModels;
            }
            tags = this.tagReadOnlyRepository.GetItemList();
            foreach (var tag in tags)
            {
                tagListModels.Add(new TagListModel
                {
                    TagId = tag.Id,
                    Text = tag.Text
                });
            }
            return tagListModels;
        }

        public TagDetailModel GetTag(int id, bool withRelated)
        {
            var tag = new Tag();
            var tagDetailModel = new TagDetailModel();
            if (withRelated)
            {
                tag = this.tagReadOnlyRepository.GetItemWithInclude(
                x => x.Id == id,
                x => x.Recipes,
                x => x.Episodes);
                tagDetailModel = new TagDetailModel
                {
                    TagId = tag.Id,
                    Text = tag.Text
                };
                return tagDetailModel;
            }
            tag = this.tagReadOnlyRepository.GetItem(id);
            tagDetailModel = new TagDetailModel
            {
                TagId = tag.Id,
                Text = tag.Text
            };
            return tagDetailModel;
        }
    }
}

File: CulinarySiteApp/src/app/services/tag.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ITagDetailModel } from '../interfaces/tag/tag-detail-model.interface';
import { ITagListModel } from '../interfaces/tag/tag-list-model.interface';
import { CreateTagModel } from '../viewmodels/tag/create-tag-model.class';
import { UpdateTagModel } from '../viewmodels/tag/update-tag-model.class';

@Injectable()
export class TagService {
  public url: string = '/api/tag';

  constructor(private http: HttpClient) {}

  public getTagList(withRelated: boolean): Observable<ITagListModel[]> {
    return this.http.get<ITagListModel[]>(this.url + '/' + withRelated);
  }

  public getTag(id: number, withRelated: boolean): Observable<ITagDetailModel> {
    return this.http.get<ITagDetailModel>(
      this.url + '/' + id + '/' + withRelated
    );
  }

  public createTag(createTagModel: CreateTagModel): Observable<void> {
    return this.http.post<void>(this.url, createTagModel);
  }

  public updateTag(updateTagModel: UpdateTagModel): Observable<void> {
    return this.http.put<void>(this.url, updateTagModel);
  }

  public deleteTag(id: number): Observable<void> {
    return this.http.delete<void>(this.url + '/' + id);
  }
}
